### PR TITLE
#19196 fix StageActorRef termination watch leak

### PR DESCRIPTION
**Pull request by rkuhn**
@drewhk One question remains: the leak actually was caused by the TcpManager actor retaining the GraphInterpreters via its watchedBy set—why is the TcpManager being watched, I wouldn’t expect it to ever terminate?

### CODE REVIEW

**Comment by rkuhn** [comment]
fixes #19196

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala [hunk 1]
```diff
@@ -138,8 +138,16 @@ object GraphStageLogic {
...
 
     /** INTERNAL API */
-    private[akka] def internalReceive(pack: (ActorRef, Any)): Unit =
-      behaviour(pack)
+    private[akka] def internalReceive(pack: (ActorRef, Any)): Unit = {
+      pack._2 match {
+        case Terminated(ref) ⇒
+          if (watching contains ref) {
+            watching -= ref
+            behaviour(pack)
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: So you *don't* want to run `behavior(pack)` is the ref is not in the watching set?

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala [hunk 2]
```diff
@@ -167,6 +175,7 @@ object GraphStageLogic {
       behaviour = receive
     }
 
+    private[this] var watching = Set.empty[ActorRef]
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: Isn't the default set a HashSet? (I'm thinking a TreeSet might be more space efficient)

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala [hunk 3]
```diff
@@ -885,8 +907,10 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
   // Internal hooks to avoid reliance on user calling super in postStop
   /** INTERNAL API */
   protected[stream] def afterPostStop(): Unit = {
-    if (_stageActorRef ne null) _stageActorRef.stop()
-    _stageActorRef = null
+    if (_stageActorRef ne null) {
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: :+1:

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala [hunk 4]
```diff
@@ -219,18 +237,22 @@ object GraphStageLogic {
...
 
-    def watch(actorRef: ActorRef): Unit =
+    def watch(actorRef: ActorRef): Unit = {
+      watching += actorRef
       actorRef.asInstanceOf[InternalActorRef].sendSystemMessage(Watch(actorRef.asInstanceOf[InternalActorRef], this))
+    }
 
-    def unwatch(actorRef: ActorRef): Unit =
+    def unwatch(actorRef: ActorRef): Unit = {
+      watching -= actorRef
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: So these can never be called concurrently?

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala (hunk 2 above)
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: Could be the cached `ActorCell.emptyActorRefSet`

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala (hunk 4 above)
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: I guess the assumption is those are called from within the Stage, so it's synchronous and safe.
// spelling it out for my understanding as well

**Comment by ktoso** [comment]
Fix makes sense, thanks!

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala (hunk 1 above)
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: Yes, these are the semantics from normal ActorCell.

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala (hunk 4 above)
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: Yes, all modifications of the `watching` set happen from within the GraphStage.

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala (hunk 2 above)
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: The default set is rather minimal for up to four entries and I switched to the cached one.

**Comment by rkuhn** [comment]
git failure …

**Comment by rkuhn** [comment]
PLS BUILD

**Comment by drewhk** [comment]
I don't think it is a bad thing to watch the TcpManager actor. If it goes away because of a bug then all listener actors will just hang around, while currently they would kill themselves failing early.

**Comment by drewhk** [comment]
LGTM otherwise

**Comment by rkuhn** [comment]
Okay, that’s some reason—better be safe than sorry (even if unlikely).

**Comment by drewhk** [comment]
I am not against removing it though if it makes things faster or eat less memory.

**Comment by rkuhn** [comment]
Nah, let’s keep the change minimal right now.

**Pull request merged**